session/txn.go

```diff
@@ -488,27 +493,46 @@ func (txn *LazyTxn) Wait(ctx context.Context, sctx sessionctx.Context) (kv.Trans
 	return txn, nil
 }
 
-func keyNeedToLock(k, v []byte, flags kv.KeyFlags) bool {
+func keyNeedToLock(k, v []byte, flags kv.KeyFlags, is infoschema.InfoSchema) (bool, error) {
 	isTableKey := bytes.HasPrefix(k, tablecodec.TablePrefix())
 	if !isTableKey {
 		// meta key always need to lock.
-		return true
+		return true, nil
 	}
 	if flags.HasPresumeKeyNotExists() {
-		return true
+		return true, nil
 	}
 
 	// lock row key, primary key and unique index for delete operation,
 	if len(v) == 0 {
-		return flags.HasNeedLocked() || tablecodec.IsRecordKey(k)
+		return flags.HasNeedLocked() || tablecodec.IsRecordKey(k), nil
 	}
 
 	if tablecodec.IsUntouchedIndexKValue(k, v) {
-		return false
+		return false, nil
 	}
-	isNonUniqueIndex := tablecodec.IsIndexKey(k) && len(v) == 1
```

Comment:
Why the old code fail to work?
For unique index, the value is encoded in the kv key part, so kv value part is not set.
For non-unique index, the value is encoded in the kv value part.
Can't `len(v) == 1` distinguish them?

Reply:
OK, get it.
In a previous refactoring, the [key-value encoding format](https://docs.google.com/document/d/1Co5iMiaxitv3okJmLYLJxZYCNChcjzswJMRr-_45Eqg/edit#heading=h.l238zyi79ds4) is changed

![image](https://user-images.githubusercontent.com/1420062/180726185-0cd89774-be49-43e9-bdcf-8fe3d28548b9.png)

Reply:
In which version we changed key-value encoding format ?

Reply:
Since the introducing of global index / clustered index / collation, whichever comes first. @TonsnakeLin